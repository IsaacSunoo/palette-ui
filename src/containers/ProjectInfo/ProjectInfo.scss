.ProjectInfo {
  display: flex;
  flex-direction: column;
  width: 50%;
  padding: 1em 2em;
}
