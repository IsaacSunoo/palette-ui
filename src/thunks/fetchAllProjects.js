import * as actions from '../actions/index';
import { fetchData } from '../utility/fetchData';
import { cleanProjectsPalettes } from '../utility/cleaner';

export const fetchAllProjects = () => {
  return async (dispatch) => {
    try {
      dispatch(actions.setLoading(true));
      const allProjects = await fetchData(process.env.REACT_APP_BACKEND_URL + '/api/v1/projects');
      const allPalettes = await fetchData(process.env.REACT_APP_BACKEND_URL + '/api/v1/palettes');
      dispatch(actions.setLoading(false));
      if(!allProjects.ok){
        throw Error(allProjects.statusText)
      }
      const combinedData = cleanProjectsPalettes(allProjects, allPalettes)
      dispatch(actions.setProjects(combinedData));
    } catch (error) {
      dispatch(actions.setError(error.message));
    }
  }
}
