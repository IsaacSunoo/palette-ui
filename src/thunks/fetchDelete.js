import * as actions from '../actions/index';
import { fetchData } from '../utility/fetchData';
import { fetchOptions } from '../utility/fetchOptions';

export const fetchDelete = (url) => {
  return async (dispatch) => {
    try {
      dispatch(actions.setLoading(true));
      const options = await fetchOptions('DELETE', {});
      await fetchData(url, options);
      dispatch(actions.setLoading(false));
    } catch (error) {
      dispatch(actions.setError(error.message));
    }
  }
}
