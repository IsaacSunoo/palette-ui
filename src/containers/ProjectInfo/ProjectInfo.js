import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import PaletteInfo from '../PaletteInfo/PaletteInfo';
import { fetchAllProjects } from '../../thunks/fetchAllProjects';
import { fetchDelete } from '../../thunks/fetchDelete';
import * as actions from '../../actions/index';

export class ProjectInfo extends Component {

  handleDelete = async (id) => {
    const url = `http://localhost:3001/api/v1/projects/${id}`;
    await this.props.fetchDelete(url);
    this.props.fetchAllProjects();
  }

  render() {
    const { name, id, palettes } = this.props;
    return (
      <div className="ProjectInfo">
        <section className="Project-Info-Header">
          <h3>{name}</h3>
          <button className="DeleteBtn" onClick={() => this.handleDelete(id)}>
            <i className="fas fa-trash-alt"></i>
          </button>
        </section>
        {
          palettes.length >= 1 ?
          palettes.map((palette, index) => {
            return <PaletteInfo key={palette.name} {...palette} id={palette.id}/>
          })
          :
          <p>Insert In Some Palettes</p>
        }
      </div>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  fetchAllProjects: (data) => dispatch(fetchAllProjects(data)),
  fetchDelete: (data) => dispatch(fetchDelete(data)),
})

ProjectInfo.propTypes = {
  allProjects: PropTypes.array,
  fetchAllProjects: PropTypes.func,
  fetchDelete: PropTypes.func,
}

export default connect(null, mapDispatchToProps)(ProjectInfo);
