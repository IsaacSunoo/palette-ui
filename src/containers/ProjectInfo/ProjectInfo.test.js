import React from 'react';
import { shallow } from 'enzyme'
import { ProjectInfo, mapDispatchToProps } from './ProjectInfo';

import { fetchAllProjects } from '../../thunks/fetchAllProjects';
import { fetchDelete } from '../../thunks/fetchDelete';
jest.mock('../../thunks/fetchAllProjects');
jest.mock('../../thunks/fetchDelete');

const mockPalettes = [
  {
    id: 15,
    name: "Mikes Pal",
    color_1: "#93125a",
    color_2: "#ffffff",
    color_3: "#f13470",
    color_4: "#80064e",
    color_5: "#423c52",
    project_id: 8,
  }
]

describe('ProjectInfo', () => {

  describe('ProjectInfo Component', () => {
    let wrapper;
    let mockfetchAllProjects = jest.fn();
    let mockFetchDelete = jest.fn();
    let mockId = 8;
    let mockName = 'Test Project Title';

    beforeEach(() => {
      wrapper = shallow(<ProjectInfo palettes={mockPalettes}
        fetchDelete={mockFetchDelete}
        fetchAllProjects={mockfetchAllProjects}
        id={mockId}
        name={mockName}
      />)
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should handleDelete on click with the correct id passed in', () => {
      const mockId = 8;
      const instance = wrapper.instance();
      const spy = jest.spyOn(instance, 'handleDelete');
      const button = wrapper.find('.DeleteBtn');
      button.simulate('click', mockId);
      expect(spy).toHaveBeenCalledWith(mockId);
    });

    it('should call fetchAllProjects when handleSendProject is invoked with correct params', async () => {
      const instance = wrapper.instance();
      const mockId = 8;
      await instance.handleDelete(mockId);
      expect(mockfetchAllProjects).toHaveBeenCalled();
    });

    it('should fetchDelete with the correct url passed in', async () => {
      const mockId = 8;
      const mockUrl = `http://localhost:3001/api/v1/projects/${mockId}`;
      const instance = wrapper.instance();
      await instance.handleDelete(mockId);
      expect(mockFetchDelete).toHaveBeenCalledWith(mockUrl);
    });

  });

  describe('mapDispatchToProps', () => {

    it('should call dispatch for fetchDelete', () => {
      const mockData = [];
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchDelete(mockData);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchDelete(mockData);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch for fetchAllProjects', () => {
      const mockData = [];
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchAllProjects(mockData);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchAllProjects(mockData);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

  });

});
