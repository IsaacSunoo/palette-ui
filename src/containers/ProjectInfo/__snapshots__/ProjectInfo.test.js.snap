// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectInfo ProjectInfo Component should match the snapshot 1`] = `
<div
  className="ProjectInfo"
>
  <section
    className="Project-Info-Header"
  >
    <h3>
      Test Project Title
    </h3>
    <button
      className="DeleteBtn"
      onClick={[Function]}
    >
      <i
        className="fas fa-trash-alt"
      />
    </button>
  </section>
  <ConnectFunction
    color_1="#93125a"
    color_2="#ffffff"
    color_3="#f13470"
    color_4="#80064e"
    color_5="#423c52"
    id={15}
    key="Mikes Pal"
    name="Mikes Pal"
    project_id={8}
  />
</div>
`;
